recipes/DebuggingLog/src/Main.purs

```diff
@@ -0,0 +1,116 @@
+module DebuggingLog.Main where
+
+import Prelude
+
+import Control.Monad.ST as ST
+import Control.Monad.ST.Internal as ST
+import Control.Parallel (parSequence)
+import Data.Int (toNumber)
+import Data.Tuple (Tuple(..))
+import Debug.Trace (spy, traceM)
+import Effect (Effect)
+import Effect.Aff (Milliseconds(..), delay, launchAff_)
+import Effect.Class (liftEffect)
+import Effect.Console (log)
+import Effect.Random (randomInt)
+
+data MyADT = MyADT Int (Tuple Int (Array String)) { foo :: String }
+
+main :: Effect Unit
+main = do
+  log "When we are in the 'Effect' monad, we can print content to the console."
+  launchAff_ do
+    liftEffect $ log $ "We can still print values to the console as long as \
+                       \ the monad in question implements the MonadEffect \
+                       \type class. Since `Aff` implements MonadEffect, we \
+                       \ can lift that effect into `Aff`."
+
+  log "However, there are times when we want to debug some code and wish \
+      \to use print-based debugging. Since PureScript is pure, how do we \
+      \do that?"
+
+  usingSpy
+
+  usingTraceMInEffect
+  usingTraceMInST
+  usingTraceMInAff
+  usingTraceMInAffRace
+
+usingSpy :: Effect Unit
+usingSpy = do
+  log "usingSpy"
+  -- `spy` returns the value it receives. However, it prints that value
+  -- to the console before it "returns."
+  --    spy :: forall a. DebugWarning => String -> a -> a
+  -- You can use this in a `let` clause pretty reliably
+  let
+    x = 5
+    y = spy "y" 8
+    adt = spy "adt" $ MyADT 1 (Tuple 4 ["a", "b"]) { foo: "foo value" }
+    function = spy "function" $ \intValue -> show $ 4 + intValue
+
+    -- quick way of debugging something without showing it or using a
+    -- variable name
+    _ = spy "debug for me what x + y is" $ x + y
+    arrayOfStrings = spy "debug some array" $ map show [1, 2, 3, 4, 5]
```

Comment:
One other place it's common to use `spy` is, for example, in a render function:

```purs
render state =
  HH.div
    [ ]
    [ HH.text state.someValue ]
```

becomes

```purs
...
    [ HH.text $ spy "state.someValue" state.someValue ]
```

This makes `spy` really convenient to drop into arbitrary code. You simply add `spy "label"` in front of whatever value it is you had before, and you get the logging without any other change to your code.

Another example: let's say you've written a little algorithm with a loop and you want to spy at each step:

```purs
fact :: Int -> Int -> Int
fact 0 acc = acc
fact n acc = fact (spy "n" (n - 1)) (spy "acc" (acc * n))
```

Being able to insert `spy` inline is really nice. I wanted to call it out because you can't do this with `traceM`.

Reply:
Added the `fact` example and updated the `traceM` usage to include two `spy`s to show how that works.